File: apps/mobile_android/test/event_gym_template_test.dart
import 'package:flutter_test/flutter_test.dart';
import '../lib/event_gym_template.dart';

void main() {
  group('parseGymTemplate', () {
    test('valid full template', () {
      final t = parseGymTemplate({'discipline': 'Vinyasa yoga', 'duration_min': 60});
      expect(t!.discipline, 'Vinyasa yoga');
      expect(t.durationMin, 60);
    });

    test('null / non-map -> null', () {
      expect(parseGymTemplate(null), isNull);
      expect(parseGymTemplate('yoga'), isNull);
      expect(parseGymTemplate(42), isNull);
      expect(parseGymTemplate(['a', 'b']), isNull);
    });

    test('empty map -> null (treated as no template)', () {
      expect(parseGymTemplate(<String, dynamic>{}), isNull);
    });

    test('discipline only', () {
      final t = parseGymTemplate({'discipline': 'Spin'});
      expect(t!.discipline, 'Spin');
      expect(t.durationMin, isNull);
    });

    test('duration only', () {
      final t = parseGymTemplate({'duration_min': 45});
      expect(t!.discipline, isNull);
      expect(t.durationMin, 45);
    });

    test('wrong-typed values are dropped', () {
      expect(parseGymTemplate({'discipline': 123, 'duration_min': 'long'}), isNull);
      expect(parseGymTemplate({'discipline': '   ', 'duration_min': 0}), isNull);
      expect(parseGymTemplate({'discipline': '  ', 'duration_min': -10}), isNull);
      final t = parseGymTemplate({'discipline': '  Pilates ', 'duration_min': 30.9});
      expect(t!.discipline, 'Pilates');
      expect(t.durationMin, 30);
    });
  });

  group('gymTemplateFromInputs', () {
    test('both empty -> null', () {
      expect(gymTemplateFromInputs('', null), isNull);
      expect(gymTemplateFromInputs('   ', null), isNull);
      expect(gymTemplateFromInputs(null, null), isNull);
      expect(gymTemplateFromInputs('', 0), isNull);
    });

    test('discipline only', () {
      final t = gymTemplateFromInputs('Strength', null);
      expect(t!.discipline, 'Strength');
      expect(t.durationMin, isNull);
    });

    test('duration only', () {
      final t = gymTemplateFromInputs('', 50);
      expect(t!.discipline, isNull);
      expect(t.durationMin, 50);
    });

    test('both, trims', () {
      final t = gymTemplateFromInputs('  Yoga  ', 60);
      expect(t!.discipline, 'Yoga');
      expect(t.durationMin, 60);
    });
  });

  group('workoutDraftFromTemplate', () {
    test('duration -> seconds, title from discipline', () {
      final d = workoutDraftFromTemplate(
        const EventGymTemplate(discipline: 'Spin', durationMin: 45),
        'Sat ride',
      );
      expect(d.title, 'Spin');
      expect(d.durationS, 45 * 60);
    });

    test('title falls back to event title when no discipline', () {
      final d = workoutDraftFromTemplate(
        const EventGymTemplate(discipline: null, durationMin: 30),
        'Morning Strength',
      );
      expect(d.title, 'Morning Strength');
      expect(d.durationS, 30 * 60);
    });

    test('null template + null title -> empty draft', () {
      final d = workoutDraftFromTemplate(null, null);
      expect(d.title, isNull);
      expect(d.durationS, isNull);
      final d2 = workoutDraftFromTemplate(null, '  ');
      expect(d2.title, isNull);
    });
  });
}

File: apps/web/src/lib/social/event_gym_template.test.ts
import { test } from 'node:test';
import assert from 'node:assert/strict';
import {
	parseGymTemplate,
	gymTemplateFromInputs,
	workoutDraftFromTemplate
} from './event_gym_template';

test('parseGymTemplate: valid full template', () => {
	assert.deepEqual(parseGymTemplate({ discipline: 'Vinyasa yoga', duration_min: 60 }), {
		discipline: 'Vinyasa yoga',
		duration_min: 60
	});
});

test('parseGymTemplate: null / non-object / array -> null', () => {
	assert.equal(parseGymTemplate(null), null);
	assert.equal(parseGymTemplate(undefined), null);
	assert.equal(parseGymTemplate('yoga'), null);
	assert.equal(parseGymTemplate(42), null);
	assert.equal(parseGymTemplate([1, 2]), null);
});

test('parseGymTemplate: empty object -> null (treated as no template)', () => {
	assert.equal(parseGymTemplate({}), null);
});

test('parseGymTemplate: discipline only', () => {
	assert.deepEqual(parseGymTemplate({ discipline: 'Spin' }), {
		discipline: 'Spin',
		duration_min: null
	});
});

test('parseGymTemplate: duration only', () => {
	assert.deepEqual(parseGymTemplate({ duration_min: 45 }), {
		discipline: null,
		duration_min: 45
	});
});

test('parseGymTemplate: wrong-typed values are dropped', () => {
	// numeric discipline + string duration + blank discipline all coerce to null
	assert.equal(parseGymTemplate({ discipline: 123, duration_min: 'long' }), null);
	assert.equal(parseGymTemplate({ discipline: '   ', duration_min: 0 }), null);
	assert.equal(parseGymTemplate({ discipline: '  ', duration_min: -10 }), null);
	assert.deepEqual(parseGymTemplate({ discipline: '  Pilates ', duration_min: 30.9 }), {
		discipline: 'Pilates',
		duration_min: 30
	});
});

test('gymTemplateFromInputs: both empty -> null', () => {
	assert.equal(gymTemplateFromInputs('', null), null);
	assert.equal(gymTemplateFromInputs('   ', null), null);
	assert.equal(gymTemplateFromInputs(null, undefined), null);
	assert.equal(gymTemplateFromInputs('', 0), null);
});

test('gymTemplateFromInputs: discipline only', () => {
	assert.deepEqual(gymTemplateFromInputs('Strength', null), {
		discipline: 'Strength',
		duration_min: null
	});
});

test('gymTemplateFromInputs: duration only', () => {
	assert.deepEqual(gymTemplateFromInputs('', 50), {
		discipline: null,
		duration_min: 50
	});
});

test('gymTemplateFromInputs: both, trims + floors', () => {
	assert.deepEqual(gymTemplateFromInputs('  Yoga  ', 60.7), {
		discipline: 'Yoga',
		duration_min: 60
	});
});

test('workoutDraftFromTemplate: duration -> seconds, title from discipline', () => {
	assert.deepEqual(workoutDraftFromTemplate({ discipline: 'Spin', duration_min: 45 }, 'Sat ride'), {
		title: 'Spin',
		duration_s: 45 * 60
	});
});

test('workoutDraftFromTemplate: title falls back to event title when no discipline', () => {
	assert.deepEqual(
		workoutDraftFromTemplate({ discipline: null, duration_min: 30 }, 'Morning Strength'),
		{ title: 'Morning Strength', duration_s: 30 * 60 }
	);
});

test('workoutDraftFromTemplate: null template + null title -> empty draft', () => {
	assert.deepEqual(workoutDraftFromTemplate(null, null), { title: null, duration_s: null });
	assert.deepEqual(workoutDraftFromTemplate(null, '  '), { title: null, duration_s: null });
});
